packages/foundry/contracts/hooks/SwapDiscountHook/DiscountCampaign.sol

```diff
@@ -0,0 +1,212 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.24;
+
+import { IERC721 } from "@openzeppelin/contracts/token/ERC721/IERC721.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+import { ReentrancyGuard } from "@openzeppelin/contracts/utils/ReentrancyGuard.sol";
+
+import { IDiscountCampaign } from "./Interfaces/IDiscountCampaign.sol";
+import { ISwapDiscountHook } from "./Interfaces/ISwapDiscountHook.sol";
+import { TransferHelper } from "./libraries/TransferHelper.sol";
+
+/**
+ * @title DiscountCampaign
+ * @notice This contract is used to manage discount campaigns, allowing users to earn rewards through swaps.
+ * @dev Implements IDiscountCampaign interface. Includes reward distribution, user discount data updates, and campaign management.
+ */
+contract DiscountCampaign is IDiscountCampaign, Ownable, ReentrancyGuard {
+    /// @notice Maps token IDs to user-specific swap data.
+    mapping(uint256 => UserSwapData) public override userDiscountMapping;
+
+    /// @notice Holds the details of the current discount campaign.
+    CampaignDetails public campaignDetails;
+
+    /// @notice Total amount of reward tokens distributed so far.
+    uint256 public tokenRewardDistributed;
+
+    /// @notice Address of the discount campaign factory that can manage campaign updates.
+    address public discountCampaignFactory;
+
+    /// @notice Address of the swap discount hook for tracking user swaps.
+    ISwapDiscountHook private _swapHook;
+
+    /// @notice Maximum buyable reward amount during the campaign.
+    uint256 private _maxBuy;
```

Comment:
This is a cap on the trade amount; i.e., rewards are proportional to the swap size, but only up to a point, after which it's flat. Maybe `swapAmountCap` or something similar?

Reply:
yes it is swapAmountCap which is related to the rewardAmount so if the swapAmount exceeds the cap then we'll calculate the reward based on the maxCap.